src/main/java/uk/gov/hmcts/ccd/domain/service/supplementarydata/DefaultSupplementaryDataUpdateOperation.java

```diff
@@ -0,0 +1,62 @@
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.BiConsumer;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.stereotype.Service;
+import uk.gov.hmcts.ccd.data.casedetails.supplementarydata.SupplementaryDataOperation;
+import uk.gov.hmcts.ccd.data.casedetails.supplementarydata.SupplementaryDataRepository;
+import uk.gov.hmcts.ccd.domain.model.std.SupplementaryData;
+import uk.gov.hmcts.ccd.domain.model.std.SupplementaryDataUpdateRequest;
+
+import static uk.gov.hmcts.ccd.data.casedetails.supplementarydata.SupplementaryDataOperation.INC;
+import static uk.gov.hmcts.ccd.data.casedetails.supplementarydata.SupplementaryDataOperation.SET;
+
+@Service
+@Qualifier("default")
+public class DefaultSupplementaryDataUpdateOperation implements SupplementaryDataUpdateOperation {
+
+    private final SupplementaryDataRepository supplementaryDataRepository;
+
+    private EnumMap<SupplementaryDataOperation, BiConsumer<String, SupplementaryDataUpdateRequest>> supplementaryFunctions =
+        new EnumMap<>(SupplementaryDataOperation.class);
+
+    @Autowired
+    public DefaultSupplementaryDataUpdateOperation(final @Qualifier("default") SupplementaryDataRepository supplementaryDataRepository) {
+        this.supplementaryDataRepository = supplementaryDataRepository;
+        supplementaryFunctions.put(SET, (caseReference, updateRequest) -> {
+            Map<String, Object> requestedData = updateRequest.getOperationProperties(SET);
+            requestedData
+                .entrySet()
+                .stream()
+                .forEach(entry -> this.supplementaryDataRepository.setSupplementaryData(caseReference,
+                    entry.getKey(),
+                    entry.getValue()));
+        });
+        supplementaryFunctions.put(INC, (caseReference, updateRequest) -> {
+            Map<String, Object> requestedData = updateRequest.getOperationProperties(INC);
+            requestedData
+                .entrySet()
+                .stream()
+                .forEach(entry -> this.supplementaryDataRepository.incrementSupplementaryData(caseReference,
+                    entry.getKey(),
+                    entry.getValue()));
+        });
+    }
+
+    @Override
+    public SupplementaryData updateSupplementaryData(String caseReference, SupplementaryDataUpdateRequest supplementaryData) {
+        supplementaryData.getRequestData().keySet().forEach(key -> {
+            Optional<SupplementaryDataOperation> operation = SupplementaryDataOperation.getOperation(key);
+            if (operation.isPresent()) {
+                supplementaryFunctions
+                    .get(operation.get())
+                    .accept(caseReference, supplementaryData);
+            }
+        });
+        return this.supplementaryDataRepository.findSupplementaryData(caseReference, supplementaryData.getRequestDataKeys());
```

Comment:
Is it possible we have a race condition here when there are many concurrent updates? 
Imagine update 1 updates propertyA: 5, then the thread stops just before line 60. At the same time update 2 updates propertyA: 10 and completes. Then update1 resumes and does the find, which will return propertyA: 10 rather than propertyA: 5. 
I think you mentioned the operations are transactional, but I can see those are transactional at the level of the Repository. Not at the level of the controller. So it's possible that update1 sees the changes introduced by update2 transaction here?

Reply:
probably better to have the repository method itself, the set, increment etch... do a find and return the updated property value. Rather than the operation.
Or move the transactionality at the level of the controller?

Reply:
@Transactional is at the controller as well!!!